src/fabric/executor/multi_threaded/executor.rs

```diff
@@ -0,0 +1,288 @@
+                if operation.inflight_args > 0 {
+                    continue;
+                }
+
+                // Mark the operation as ready for execution
+                ready_ops.push(*op_id);
+            }
+        }
+
+        self.execute_operations(ready_ops);
+    }
+
+    /// Executes the operations in the buffer, recursively executing any
+    /// dependencies that become ready
+    fn execute_operations(&mut self, ops: Vec<OperationId>) {
+        for op_id in ops.into_iter() {
+            let op = self.operations.take(op_id).unwrap();
+
+            let results = self.results.clone();
+            let job_queue = self.job_queue.clone();
+            let network_outbound = self.network_outbound.clone();
+            self.pool.spawn(move || Self::compute_result(op, results, job_queue, network_outbound));
+        }
+    }
+
+    /// Compute the result of an operation, assumed to be done in a separate
+    /// thread that will re-enqueue the result with the coordinator
+    fn compute_result(
+        op: Operation<C>,
+        result_buffer: ParallelResultBuffer<C>,
+        job_queue: ExecutorJobQueue<C>,
+        network_sender: KanalSender<NetworkOutbound<C>>,
+    ) {
+        let result_ids = op.result_ids();
+
+        // Collect the inputs to the operation
+        let args = op.args.into_iter().map(|arg| result_buffer.get(arg).unwrap().clone());
+        let input = Box::new(args);
+
+        let results = match op.op_type {
+            OperationType::Gate { function } => {
+                let value = (function)(input);
+                vec![OpResult { id: op.result_id, value }]
+            },
+
+            OperationType::GateBatch { function } => {
+                let output = (function)(input);
+                result_ids
+                    .into_iter()
+                    .zip(output)
+                    .map(|(id, value)| OpResult { id, value })
+                    .collect()
+            },
+
+            OperationType::Network { function } => {
+                // Derive a network payload from the gate inputs and forward it to the outbound
+                // buffer
+                let result_id = result_ids[0];
+                let payload = (function)(input);
+                let outbound = NetworkOutbound { result_id, payload: payload.clone() };
+
+                network_sender.send(outbound).expect("error sending network payload");
+
+                // On a `send`, the local party receives a copy of the value placed as the
+                // result of the network operation, so we must re-enqueue the
+                // result
+                vec![OpResult { id: result_id, value: payload.into() }]
+            },
+        };
+
+        // Place the results in the result buffer
+        let mut ids = Vec::with_capacity(results.len());
+        for result in results.into_iter() {
+            ids.push(result.id);
+            result_buffer.set(result.id, result.value);
```

Comment:
should we also check for the result ID already existing in the result buffer as we do in `insert_result`?